tests/cache_tests/test_cache.py

```diff
@@ -3,105 +3,133 @@
 import pickle
 import random
 import tempfile
+from contextlib import contextmanager
+from threading import Thread
 
 import numpy as np
 import pytest
 
-from pfio.cache import FileCache, MultiprocessFileCache, NaiveCache
+from pfio.cache import FileCache, MultiprocessFileCache, NaiveCache, HTTPCache
+from pfio.testing import make_http_server
 
 
+@contextmanager
 def make_cache(test_class, mt_safe, do_pickle, length,
                cache_size_limit=None):
-    if test_class == NaiveCache:
-        assert cache_size_limit is None, \
-            "NaiveCache doesn't support cache size limit"
-        cache = test_class(length, multithread_safe=mt_safe,
-                           do_pickle=do_pickle)
-        assert not cache.multiprocess_safe
-        assert cache.multithread_safe == mt_safe
+    httpd = None
+    httpd_thread = None
 
-    elif test_class == FileCache:
-        cache = test_class(length, multithread_safe=mt_safe,
-                           cache_size_limit=cache_size_limit,
-                           do_pickle=do_pickle)
-        assert not cache.multiprocess_safe
-        assert cache.multithread_safe == mt_safe
-
-    elif test_class == MultiprocessFileCache:
-        cache = test_class(length, cache_size_limit=cache_size_limit,
-                           do_pickle=do_pickle,)
-        assert cache.multiprocess_safe
-        assert cache.multithread_safe
-
-    else:
-        assert False
-
-    return cache
+    try:
```

Comment:
This try..cache clause is only for HTTP Cache and much wider than the actual scope it needs to be in. To clarify it, please use it inside the `elif test_class == HTTPCache:` clause, e.g.

```python
        elif test_class == HTTPCache:
            assert cache_size_limit is None, \
                "HTTPCache doesn't support cache size limit"

            try:
              httpd, port = make_http_server()
              httpd_thread = Thread(target=httpd.serve_forever)
              httpd_thread.start()
            
              cache = test_class(length, f"http://localhost:{port}/", do_pickle=do_pickle)
              assert cache.multiprocess_safe
              assert cache.multithread_safe
              yield cache
            finally:
              ...
```

Another option to clarify the code is to make the `make_http_server()` as context manager like this:

```python
@contextmanager
def new_make_http_server():
    try:
       httpd, port = make_http_server()
       httpd_thread = Thread(target=httpd.serve_forever)
       httpd_thread.start()
       yield f"http://localhost:{port}/"
    finally:
       ...
```

Then the test would be more simple without large try-cache clause, like
```python
        elif test_class == HTTPCache:
            assert cache_size_limit is None, \
                "HTTPCache doesn't support cache size limit"

            with new_make_cache() as cache_server_url:
              cache = test_class(length, f"http://localhost:{port}/", do_pickle=do_pickle)
              assert cache.multiprocess_safe
              assert cache.multithread_safe
              yield cache
```

Reply:
I've added `contextmanager` in `make_http_server` to make sure server shutdown will be performed by https://github.com/pfnet/pfio/pull/282/commits/a8e576c484625b8d095004980a0b4a3d564945db. By this, I can avoid using try-catch clause in `make_cache`.